### PR TITLE
Read package.json from APP_ROOT or Rails.root, not cwd

### LINKED ISSUE #519: error_unless_package_manager_is_obvious! triggers when rails environment is required from outside the rails root
## Expected behavior:

When running rails code from outside the rails root, it should still work.

## Actual behavior:

It errors because of the package json check:
```
vendor/bundle/ruby/3.3/gems/package_json-0.1.0/lib/package_json.rb:25:in `read': $DIR does not contain a package.json (PackageJson::Error)
from vendor/bundle/ruby/3.3/gems/shakapacker-8.0.2/lib/shakapacker/utils/manager.rb:19:in `error_unless_package_manager_is_obvious!'
from vendor/bundle/ruby/3.3/gems/shakapacker-8.0.2/lib/shakapacker/railtie.rb:20:in `block in <class:Engine>'
from vendor/bundle/ruby/3.3/gems/railties-7.1.3.4/lib/rails/initializable.rb:32:in `instance_exec'
from vendor/bundle/ruby/3.3/gems/railties-7.1.3.4/lib/rails/initializable.rb:32:in `run'
from vendor/bundle/ruby/3.3/gems/railties-7.1.3.4/lib/rails/initializable.rb:61:in `block in run_initializers'
from /usr/local/lib/ruby/3.3/tsort.rb:231:in `block in tsort_each'
from /usr/local/lib/ruby/3.3/tsort.rb:353:in `block (2 levels) in each_strongly_connected_component'
from /usr/local/lib/ruby/3.3/tsort.rb:434:in `each_strongly_connected_component_from'
from /usr/local/lib/ruby/3.3/tsort.rb:352:in `block in each_strongly_connected_component'
from /usr/local/lib/ruby/3.3/tsort.rb:350:in `each'
from /usr/local/lib/ruby/3.3/tsort.rb:350:in `call'
from /usr/local/lib/ruby/3.3/tsort.rb:350:in `each_strongly_connected_component'
from /usr/local/lib/ruby/3.3/tsort.rb:229:in `tsort_each'
from /usr/local/lib/ruby/3.3/tsort.rb:208:in `tsort_each'
from vendor/bundle/ruby/3.3/gems/railties-7.1.3.4/lib/rails/initializable.rb:60:in `run_initializers'
from vendor/bundle/ruby/3.3/gems/railties-7.1.3.4/lib/rails/application.rb:426:in `initialize!'
from config/environment.rb:12:in `<top (required)>'
from <internal:/usr/local/lib/ruby/3.3/rubygems/core_ext/kernel_require.rb>:136:in `require'
from <internal:/usr/local/lib/ruby/3.3/rubygems/core_ext/kernel_require.rb>:136:in `require'
from test/test_helper.rb:25:in `<top (required)>'
```

test_helper.rb:25 contains

```
require File.expand_path('../config/environment', __dir__)
```

## Small, reproducible repo:

Should be easy to set up, just init a rails repo, lets say in "reproducer", then add a ruby file one level up with the following content:

```ruby
require File.expand_path('reproducer/config/environment', __dir__)
```

## Setup environment:
- Ruby version: 3.3.2
- Rails version: 7.1.3.4
- Shakapacker version: 8.0.2

I think the issue is in 

```ruby
      def self.error_unless_package_manager_is_obvious!
        return unless PackageJson.read.fetch("packageManager", nil).nil?
```

This should be something along the lines of
```ruby
      def self.error_unless_package_manager_is_obvious!
        return unless PackageJson.read(Rails.root).fetch("packageManager", nil).nil?
```

(at least this fixes it for me - of course you might not what the dependency to rails here, but rather in the railtie..).

Btw. self.guess_binary seems to have the same issue, it looks for the lockfile in the current directory instead of the rails root.
On the other hand, Shakapacker::VersionChecker::NodePackageVersion.package_json_path finds the correct package.json, so maybe error_unless_package_manager_is_obvious should use that instead?

### PR DESCRIPTION
**Pull request by marvinthepa**
Fixes #519.

### Summary

We have a few tools that are called from outside the rails root, require the rails environment, and then do something with the rails db.
These fail as error_unless_package_manager_is_obvious only works when called from inside Rails.root.
Other methods like `Shakapacker::VersionChecker::NodePackageVersion.package_json_path` take this into account, so my guess is that this is just a small bug.

### Pull Request checklist

- [x] Add/update test to cover these changes
- [x] ~Update documentation~ Bugfix, not a documented feature
- [x] Update CHANGELOG file

## Summary by CodeRabbit

- **Documentation**
	- Updated the changelog to include versioning, upgrade guides, and significant changes across multiple versions of the Shakapacker library.

- **New Features**
	- Enhanced package manager detection to operate within the correct file path context of a Rails application.

- **Bug Fixes**
	- Addressed errors related to the Rails environment being required from outside the Rails root directory.
	- Fixed documentation for the esbuild loader and improved asset compilation issues.

- **Tests**
	- Expanded test coverage for scenarios involving different working directories and package manager detection logic.

### CODE REVIEW

File: lib/shakapacker/utils/manager.rb [hunk 1]
```diff
@@ -16,7 +16,7 @@ class Error < StandardError; end
 
       # Emits a warning if it's not obvious what package manager to use
       def self.error_unless_package_manager_is_obvious!
-        return unless PackageJson.read.fetch("packageManager", nil).nil?
+        return unless PackageJson.read(Rails.root).fetch("packageManager", nil).nil?
```
[the remark this answers is not in the document]
Comment (marvinthepa) [reply lib/shakapacker/utils/manager.rb:19]: `lib/shakapacker/runner.rb` already uses

```ruby
PackageJson.read(@app_path)
```

in `lib/install/template.rb`, the call is wrapped inside

```ruby
Dir.chdir(Rails.root) do  
```

File: spec/shakapacker/utils_manager_spec.rb [hunk 2]
```diff
@@ -63,10 +64,35 @@ def exitstatus
...
+
+          expect { Shakapacker::Utils::Manager.error_unless_package_manager_is_obvious! }.to raise_error(Shakapacker::Utils::Manager::Error, <<~MSG)
+            You don't have "packageManager" set in your package.json
+            meaning that Shakapacker will use npm but you've got a #{lock}
+            file meaning you probably want to be using #{manager} instead.
+
+            To make this happen, set "packageManager" in your package.json to #{manager}@1.2.3
+          MSG
+        end
+      end
```
[the remark this answers is not in the document]
Comment (marvinthepa) [reply spec/shakapacker/utils_manager_spec.rb:89]: > is duplicated for each package manager

this code is not duplicated for each package manager.

File: spec/shakapacker/utils_manager_spec.rb [hunk 3]
```diff
@@ -91,7 +130,25 @@ def exitstatus
...
+          allow(Rails).to receive(:root).and_return(Pathname.new("."))
+        end
+
+        it "calls #{manager} with --version" do
+          Shakapacker::Utils::Manager.guess_version
+
+          expect(Open3).to have_received(:capture3).with("#{manager} --version")
+        end
+      end
+
```
[the remark this answers is not in the document]
Comment (marvinthepa) [reply spec/shakapacker/utils_manager_spec.rb:144]: duplicate to next comment

File: spec/shakapacker/utils_manager_spec.rb [hunk 4]
```diff
@@ -75,6 +101,19 @@ def exitstatus
...
+          rails_root = Pathname.new("rails_root_#{lock}")
+          FileUtils.mkdir_p(rails_root)
+          FileUtils.touch(rails_root.join(lock))
+          allow(Rails).to receive(:root).and_return(rails_root)
+        end
+
+        it "guesses #{manager}" do
+          expect(Shakapacker::Utils::Manager.guess_binary).to eq manager
+        end
+      end
```
[the remark this answers is not in the document]
Comment (marvinthepa) [reply spec/shakapacker/utils_manager_spec.rb:116]: error_unless_package_manager_is_obvious is not a test method, so extracting into a shared context won't help here.

File: spec/shakapacker/utils_manager_spec.rb [hunk 5]
```diff
@@ -91,7 +130,25 @@ def exitstatus
...
+
+      context "when lockfile is in Rails.root, but pwd is different" do
+        before do
+          rails_root = Pathname.new("rails_root_#{lock}")
+          FileUtils.mkdir_p(rails_root)
+          FileUtils.touch(rails_root.join(lock))
+          allow(Rails).to receive(:root).and_return(rails_root)
+        end
```
[the remark this answers is not in the document]
Comment (marvinthepa) [reply spec/shakapacker/utils_manager_spec.rb:151]: If a human reviewer wants me to extract a shared context for something as simple as this, I am more than willing to oblige.

IMVHO, tests profit from a little bit of duplication if it makes the logic easier to follow. Especially rspec shared contexts sometimes make tests unecessarily complicated.

**Review by justin808 — APPROVED** [review]

**Review by G-Rath — APPROVED** [review]
yyeaah this sounds reasonable to me - if nothing else, I know the context and where to look if I later find an edge case 😄 

nice work on the tests btw!

**Comment by justin808** [comment]
Once specs pass, I'll merge.

**Comment by justin808** [comment]
@marvinthepa got a test failure.

**Comment by marvinthepa** [comment]
> @marvinthepa got a test failure.

Fixed.

In the shakapacker binstub, `Rails.root` is not available, so I am using APP_ROOT there. I don't know if that is a good solution, I would be glad to get any suggestions.

**Comment by justin808** [comment]
> > @marvinthepa got a test failure.
> 
> Fixed.
> 
> In the shakapacker binstub, `Rails.root` is not available, so I am using APP_ROOT there. I don't know if that is a good solution, I would be glad to get any suggestions.

where did you get APP_ROOT?

https://chatgpt.com/share/67085053-11d0-8002-b73a-e337730b805d

**Comment by marvinthepa** [comment]
> > In the shakapacker binstub, `Rails.root` is not available, so I am using APP_ROOT there. I don't know if that is a good solution, I would be glad to get any suggestions.

> where did you get APP_ROOT?

it is  defined in the binstubs:

https://github.com/shakacode/shakapacker/blob/9544d5521488aae368ab2c61b8dcb5fa3c419445/lib/install/bin/shakapacker#L10
https://github.com/shakacode/shakapacker/blob/9544d5521488aae368ab2c61b8dcb5fa3c419445/lib/install/bin/shakapacker-dev-server#L10

**Comment by alecslupu** [comment]
@justin808 any plans to release this fix anytime soon ?

**Comment by alecslupu** [comment]
> @justin808 any plans to release this fix anytime soon ?

@justin808 , @G-Rath  any news ?

**Comment by justin808** [comment]
I will push a release today.

**Comment by alecslupu** [comment]
> I will push a release today.

Looking forward to upgrade

**Comment by justin808** [comment]
Shipped 8.1.0!
